python/lsst/daf/butler/core/dimensions/sets.py

```diff
@@ -0,0 +1,558 @@
+    accept other set-like objects and iterables containing either
+    `DimensionElement` instances or their string names; because
+    `DimensionElement`\s cannot be directly constructed, APIs that accept them
+    should generally accept a name as an alternative when the transformation
+    to a `DimensionElement` can be done internally.  Operators that return
+    new sets (`|`, `&`, `^`, and `^`) do require `DimensionSet` operands on
+    both sides to avoid surprises in return types.
+
+    Because the `DimensionElement` objects they hold always have a name,
+    `DimensionSet`\s also supports some `dict`-like operations: including
+    regular square-bracket indexing (`__getitem__`), `get`, and the ``in``
+    operator (`__contains__`).  Both names and `DimensionElement` objects can
+    be passed to any of these. The `names` attribute can also be used to
+    obtain a `set`-like object containing those names.
+
+    `DimensionSet` instances cannot be constructed directly; they can only be
+    obtained (possibly indirectly) from a special "universe" `DimensionGraph`
+    loaded from configuration.
+
+    Parameters
+    ----------
+    univsere : `DimensionGraph`
+        Ultimate-parent `DimensionGraph` that constructed the elements in
+        this set.
+    elements : iterable of `DimensionElement` or `str`
+        Elements to include in the set, or names thereof.
+    expand : `bool`
+        If `True`, recursively expand the set to include dependencies.
+    optional : `bool`
+        If `True`, include optional dependencies in expansion.  Ignored
+        if ``expand`` is `False`.
+
+    """
+    def __init__(self, universe, elements, expand=False, optional=False):
+        self._universe = universe
+        self._elements = OrderedDict()
+
+        def toPairs(elems):
+            for elem in elems:
+                if not isinstance(elem, DimensionElement):
+                    elem = self._universe.elements[elem]
+                yield (elem.name, elem)
+                if expand:
+                    yield from toPairs(elem.dependencies(optional=optional))
+
+        names = dict(toPairs(elements))
+        if names:
+            for candidate in self._universe.elements:
+                if names.pop(candidate.name, None):
+                    self._elements[candidate.name] = candidate
+
+    @property
+    def universe(self):
+        """The graph of all dimensions compatible with self (`DimensionGraph`).
+        """
+        return self._universe
+
+    def __len__(self):
+        return len(self._elements)
+
+    def __iter__(self):
+        return iter(self._elements.values())
+
+    def __str__(self):
+        return "{{{}}}".format(", ".join(self._elements.keys()))
+
+    def __repr__(self):
+        return f"DimensionSet({self})"
+
+    def __contains__(self, key):
+        try:
+            return key.name in self._elements
+        except AttributeError:
+            return key in self._elements
```

Comment:
I certainly strongly agree with that principle in C++; in Python I also prefer to avoid exceptions in control-flow but don't think it's worth much effort to avoid them (i.e. if the alternative involves a big `if` or a lot of otherwise-unnecessary variables).  But your alternative in this case is clearly better; will do.